File: components/auth/src/main/java/org/cloudfoundry/credhub/interceptors/UserContextInterceptor.java
package org.cloudfoundry.credhub.interceptors;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;

import org.cloudfoundry.credhub.auth.UserContextFactory;
import org.cloudfoundry.credhub.auth.UserContextHolder;

@Component
public class UserContextInterceptor extends HandlerInterceptorAdapter {
  private final UserContextFactory userContextFactory;
  private final UserContextHolder userContextHolder;

  @Autowired
  UserContextInterceptor(
    final UserContextFactory userContextFactory,
    final UserContextHolder userContextHolder) {
    super();
    this.userContextFactory = userContextFactory;
    this.userContextHolder = userContextHolder;
  }

  @Override
  public boolean preHandle(final HttpServletRequest request, final HttpServletResponse response, final Object handler) {
    final Authentication principal = (Authentication) request.getUserPrincipal();
    if (principal == null) {
      return false;
    }
    userContextHolder.setUserContext(userContextFactory.createUserContext(principal));
    return true;
  }
}

File: components/auth/src/main/kotlin/org/cloudfoundry/credhub/interceptors/UserContextInterceptor.kt
package org.cloudfoundry.credhub.interceptors

import org.cloudfoundry.credhub.auth.UserContextFactory
import org.cloudfoundry.credhub.auth.UserContextHolder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.servlet.HandlerInterceptor

@Component
class UserContextInterceptor
    @Autowired
    internal constructor(
        private val userContextFactory: UserContextFactory,
        private val userContextHolder: UserContextHolder,
    ) : HandlerInterceptor {
        override fun preHandle(
            request: HttpServletRequest,
            response: HttpServletResponse,
            handler: Any,
        ): Boolean {
            val principal =
                if (request.userPrincipal != null) {
                    request.userPrincipal as Authentication
                } else {
                    return false
                }
            userContextHolder.userContext = userContextFactory.createUserContext(principal)
            return true
        }
    }
